crates/wasm-shrink/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+            self.predicate.is_file(),
+            "The predicate script '{}' does not exist.",
+            self.predicate.display()
+        );
+        anyhow::ensure!(
+            self.predicate.is_executable(),
+            "The predicate script '{}' is not executable.",
+            self.predicate.display()
+        );
+
+        // Prerequisites for the input Wasm.
+        anyhow::ensure!(
+            self.input.is_file(),
+            "The input Wasm file '{}' does not exist.",
+            self.input.display()
+        );
+        anyhow::ensure!(
+            self.is_valid_wasm(&self.input)?,
+            "The input Wasm file '{}' is not valid Wasm.",
+            self.input.display(),
+        );
+
+        let input_size = self.input.metadata()?.len();
+
+        let output = self
+            .output
+            .clone()
+            .unwrap_or_else(|| self.input.with_extension("shrunken.wasm"));
+        log::info!("Writing shrunken Wasm file to: {}", output.display());
+
+        // First double check that the input Wasm file passes the predicate.
+        //
+        // It is fairly common to accidentally write a predicate script that
+        // doesn't consider the input Wasm interesting. Better to surface this
+        // user error as quick as possible than to make them wait until we've
+        // exhausted all the ways we could shrink it further.
+        anyhow::ensure!(
+            self.test_file(&self.input)?,
+            "The predicate does not consider the input Wasm file interesting."
+        );
+
+        // The smallest Wasm that passes the predicate.
+        let mut best = vec![];
+        read_file(&self.input, &mut best)?;
+
+        let new_best = |old_best: &mut Vec<u8>, new: Vec<u8>| -> Result<()> {
+            debug_assert!(
+                new.len() < old_best.len() || (new == EMPTY_WASM && *old_best == EMPTY_WASM)
+            );
+            write_file(&output, &new)?;
+            log::info!(
+                "New smallest interesting Wasm file discovered: {} bytes ({:.02}% smaller)",
+                new.len(),
+                (100.0 - (new.len() as f64 / input_size as f64 * 100.0))
+            );
+            *old_best = new;
+            Ok(())
+        };
+
+        // Next try running the predicate on an empty Wasm module.
+        //
+        // It is fairly common to accidentally write a predicate script that
+        // considers the empty module interesting, and we might as well check
+        // for it eagerly, rather than make the user wait forever until we
+        // finally to reduce the whole Wasm module to nothing.
+        if self.test_bytes(EMPTY_WASM)? {
+            new_best(&mut best, EMPTY_WASM.to_vec())?;
+            log::warn!(
+                "The predicate considers the empty Wasm module \
+                 interesting. This is usually not desired and \
+                 is a symptom of a bug in the predicate script."
+            );
+        } else {
```

Comment:
Could this `else` simply follow the `if` above unconditionally? (avoiding extra indentation)